### PR TITLE
feat(cache): Add SpeckleCache

### PR DESCRIPTION
**Pull request by tsvilans**
### SpeckleCache

* Adds separate SpeckleCache class for managing the local database (`Cache.py`).
* Cleans up database operations in`SpeckleApiClient`.
* Removes older profile loading methods.

BREAKING CHANGE: move database functions from SpeckleClient to SpeckleCache
    
    To migrate the code follow the example below:
    
    Before:
    
    client = SpeckleApiClient()
    accounts = client.load_local_profiles_from_database()
    
    profile = {'email':'foo@bar.com', 'server':'https://hestia.speckle.works/api/v1', 'server_name':'Hestia Speckle', 'apitoken':'abcdefgh'}
    client.write_profile_to_database(profile)
    
    After:

    cache = SpeckleCache()
    accounts = cache.get_all_accounts()

    cache .write_account("https://hestia.speckle.works/api/v1", "Hestia Speckle", "foo@bar.com", "abcdefgh")
    
    The object and user profile database should be independent of the API clients.

### CODE REVIEW

File: speckle/SpeckleClient.py
Comment (AntoineDao) [inline speckle/SpeckleClient.py:37]: That looks like a breaking change to me. Probably worth notifying it in the commit.

File: speckle/resources/objects.py
Comment (AntoineDao) [inline speckle/resources/objects.py]: What's going on here? How come this counts as a change when the master branch already has this? Would you mind refreshing your changes from the upstream master branch and recommitting so these non-changes don't appear here?

File: speckle/Cache.py
Comment (AntoineDao) [inline speckle/Cache.py:27]: Neat new feature 😄 Can I convince you to write tests for these so that if we make changes later we're sure to not break things? 

Also worth adding some documentation for this no?

**Review by AntoineDao — CHANGES_REQUESTED** [review]
Would be nice to add some documentation and pull the upstream master to your local branch before recommitting your changes to clean things up (but this is not essential).

However it would be great if we could do two things before we merge this:
1. Write some tests for your new code
2. Recommit your changes using [Angular commit message format](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#format-of-the-commit-message) and especially notifying that you have introduced [breaking changes](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#breaking-changes) in the footer of the commit message so `semantic-release` knows to bump by a whole version.

File: speckle/resources/objects.py
Comment (tsvilans) [reply speckle/resources/objects.py]: Yeps, sorry, I've merged the latest changes from upstream master.

File: speckle/Cache.py
Comment (tsvilans) [reply speckle/Cache.py:27]: Yes, this is quite preliminary, I guess a bit too eager to get it out the door. I'll try to make some time to write some tests + docs ;)

Let's hold off on this PR for now then.

File: speckle/resources/objects.py
Comment (AntoineDao) [reply speckle/resources/objects.py]: Yeah that works, commented on the PR you made but it's just me being picky: https://github.com/speckleworks/PySpeckle/pull/20#issuecomment-518197590

File: speckle/Cache.py
Comment (AntoineDao) [reply speckle/Cache.py:27]: Well we don't have to hold it off. You can take your time to write tests and once you're ready just ping me to review it again 😄

**Comment by AntoineDao** [comment]
Yo @tsvilans I assume you're the one making the last couple of changes. It's worth adding info to your local git profile on command line (if that's what you're using) so we can recognise you as the committer 😄 
You can do so by running these two commands in your terminal.
> git config --global user.name "Tom Svilans"
> git config --global user.email "tom.svilans@gmail.com"

PS: nice commit messages 😉

File: speckle/Cache.py
Comment (AntoineDao) [inline speckle/Cache.py:37]: Does this also work on Unix systems?

**Review by AntoineDao — COMMENTED** [review]
Not a biggie as this PR contains a feature change but it's worth noting that documention changes should be in format: `docs(cache): add inline api doc`.

Full list of change types can be found [here](https://gist.github.com/brianclements/841ea7bffdb01346392c#type)

File: speckle/Cache.py
Comment (tsvilans) [reply speckle/Cache.py:37]: Not sure... but needs to be wherever Speckle puts `SpeckleCache.db`.

I guess we should do an OS check for the default database path.

@didimitrie Any suggestions for where to stick the cache on Unix systems?

**Comment by tsvilans** [comment]
> Not a biggie as this PR contains a feature change but it's worth noting that documention changes should be in format: `docs(cache): add inline api doc`.
> 
> Full list of change types can be found [here](https://gist.github.com/brianclements/841ea7bffdb01346392c#type)

Cheers, still getting to grips with the commit style!

**Comment by tsvilans** [comment]
@AntoineDao I'm looking at how to write a test for `SpeckleCache`... should this be part of this PR or done separately? If so, and there are no further issues, then I could merge.

File: speckle/Cache.py
Comment (AntoineDao) [reply speckle/Cache.py:37]: I guess the question was also where the local app data variable is set? On unix you can usually dump it in something like `~/.speckle/`

File: speckle/SpeckleClient.py
Comment (tsvilans) [reply speckle/SpeckleClient.py:37]: I changed the PR comments to include BREAKING CHANGE. Hope I did it correctly...

File: speckle/Cache.py
Comment (tsvilans) [reply speckle/Cache.py:37]: %LOCALAPPDATA% is a Windows variable. I'll change it to check the OS, and set it accordingly.

**Comment by tsvilans** [comment]
> Yo @tsvilans I assume you're the one making the last couple of changes. It's worth adding info to your local git profile on command line (if that's what you're using) so we can recognise you as the committer 😄
> You can do so by running these two commands in your terminal.
> 
> > git config --global user.name "Tom Svilans"
> > git config --global user.email "[tom.svilans@gmail.com](mailto:tom.svilans@gmail.com)"
> 
> PS: nice commit messages 😉

Yep, that's me ;) Updated my login info, it wasn't set on this computer :P

Re: commit messages, slowly learning...

File: speckle/Cache.py
Comment (AntoineDao) [inline speckle/Cache.py]: Why is there a __main__ method here? Are we likely to run this code directly from file? Also, why are we creating dummy accounts by default?

File: speckle/SpeckleClient.py
Comment (AntoineDao) [reply speckle/SpeckleClient.py:37]: PR comment won't affect the semantic-release process unfortunately. Only the commit messages count. However we could squash and merge this to make it work. Or, even easier, you could install commitizen on your machine and create commits using it. You'll need to install nodejs and then do the following:

* `npm install -g commitizen cz-conventional-changelog`
* go to your pyspeckle folder
* make some changes
* `git add .`
* `git cz and follow the command line prompt`

Trust me, this will make your life so much easier in terms of this commit format.

File: speckle/Cache.py
Comment (tsvilans) [reply speckle/Cache.py]: **main()** is just for testing, when `Cache.py` is executed by itself. Can be removed, no biggie.

Comment (AntoineDao) [reply speckle/Cache.py]: If it's just for testing I reckon we should dump the code in the `tests` folder in a file called `test_cache.py`. Then you can use standard `pytest` to write your tests and make sure it all works. To be honest I'm not sure what all this local db cache is for so I trust you to write the appropriate tests for it 😉

**Comment by AntoineDao** [comment]
> @AntoineDao I'm looking at how to write a test for `SpeckleCache`... should this be part of this PR or done separately? If so, and there are no further issues, then I could merge.

I personally would think it's a better idea. However if you're happy committing untested code (I'm intentionally making it sound bad 😛 ) then go ahead. Let's make sure that you insert a commit with `BREAKING CHANGE: blablabla` at the footer of a commit message somewhere.

File: tests/resources/test_cache.py
Comment (AntoineDao) [inline tests/resources/test_cache.py]: Remove this line. You don't need it.

**Pull request merged**
